cmd/kubeadm/app/cmd/phases/join/kubelet.go

```diff
@@ -128,6 +130,39 @@ func runKubeletStartJoinPhase(c workflow.RunData) (returnErr error) {
 		return errors.Errorf("couldn't create client from kubeconfig file %q", bootstrapKubeConfigFile)
 	}
 
+	// Make sure to exit before TLS bootstrap if a node with the same name exist in the cluster.
+	// A new node with the same name as an existing control-plane node can cause undefined
+	// behavior and ultimatelly control-plane failure.
+	//
+	// Given there are multiple ways to define the node name we apply the following precedence
+	// from lower to higher:
+	// - actual hostname
+	// - JoinConfiguration.NodeRegistration.Name (same as "--node-name" passed to "kubeadm join")
+	// - "hostname-overide" flag in JoinConfiguration.NodeRegistration.KubeletExtraArgs
+	//
+	// The same precedence already applies in app/phases/kubelet/flags.go
```

Comment:
can the logic be shared? two places doing this fallback are likely to drift

Reply:
it's slightly more difficult than it seems and will extend the scope of this PR (i.e. include refactors), but i can give it a shot.